### DISCUSSION TITLE
C-x and KEYPAD x look wildly different

### DISCUSSION THREAD

**Discussion by snowiow** [meow-edit/meow#226] [Q&A] [answered]
Hi, I'm currently checking out meow and overall I'm really astonished how polished this mode is. 

I have a question: Somehow `C-x` and KEYPAD x yield wildly different outputs. If I understood KEYPAD mode correctly, pressing x afterwards should be the same as pressing `C-x`.  However the follow up keys shown by which-key are wildly different. I have attached screenshots for both here:

C-x:
![Screenshot 2022-04-28 at 13 19 54](https://user-images.githubusercontent.com/3718461/165741174-e057324a-2e04-4e0f-b94d-c0d63504c9c3.png)

KEYPAD x:
![Screenshot 2022-04-28 at 13 20 05](https://user-images.githubusercontent.com/3718461/165741272-237320ce-be4f-4a8e-bf29-e7cac3fd7f7e.png)

Am I doing something wrong or did I misunderstood something about KEYPAD mode. Thanks in advance for your help 🙂

**Comment by eshrh** [comment, marked as answer]
If you're wondering about the difference in actual keys shown, its because keypad C-x will only show you keys that begin with control (when you press `SPC x f` you are calling `C-x C-f`. If you press space, you'll see only the keys that don't begin with a modifier `SPC x SPC f` is `C-x f`.

As for the color difference, formatting, and pretty symbols; i'm not sure but I can look into it

**Reply by snowiow** [reply]
Ah thanks, that totally makes sense now. Thanks for clarifying!

**Reply by aisamu** [reply]
That makes sense, thanks!

I don't mind the color difference, but would be very useful to fix the group renaming (`useful-name` -> `+prefix`)

Since this is happening "frequently" (n++ https://github.com/meow-edit/meow/issues/337#event-7528441577), would it be possible to add a workaround?

Using the example above, I think it'd be helpful to show the missing entries with a `○` prefix (or something like it), as in: 
>  `○-SPC backward-kill-sentence`
>  `○-ESC repeat-complex-command`

**Reply by eshrh** [reply]
You're right, something should be done.

I think a good idea could be to instead change the display format of your currently entered keys.
For instance,

Typing `SPC x` prints `C-x C-○`
Typing `SPC x SPC` prints `C-x ○`

I feel like showing both keybinds under control and plain keybinds in the popup window is a bit cluttered.

**Reply by DogLooksGood** [reply]
Well, Emacs itself has a way to name groups. [Check this example](https://github.com/emacs-mirror/emacs/blob/3af9e84ff59811734dcbb5d55e04e1fdb7051e77/lisp/kmacro.el#L203)

And meow follows as much Emacs does as possible, so the alias can be handled well. If you press, `SPC x`, you will see `+kmacro-keymap`. But unfortunately, there are some other ways to set description for groups, provided by packages.

**Reply by aisamu** [reply]
> change the display format of your currently entered keys.

Would you think it's fair to say the "spirit" of `which-keys` revolves around discoverability? 
Even if we change the display format to be more accurate, nesting multiple keys behind a `SPC` group might not be _truly_ helping the user on that front. [^1]

There's also some value in being less surprising even if meow is 100% correct (as evidenced by the number of different issues filed on this same subject)

> I feel like showing both keybinds under control and plain keybinds in the popup window is a bit cluttered.

I 100% agree that too many keys shown may be overwhelming. 
In this case, it doesn't look like we would add/remove entries compared to the "regular" view:
* `C-;` -> `;`
* `SPC` -> `○-SPC`

We could even argue that some would become less cluttered 😄      

NB: not arguing against/for this specific implementation, just the general idea!

[^1]: Speaking from experience 😅 - I come from evil and have no idea what the "regular" emacs bindings are. I usually `M-x` to find the command keys and take the opportunity to learn a couple sibling bindings via the `which-key` popup.

**Reply by DogLooksGood** [reply]
You can use `which-key` with meow. The default popup shipped by meow is just a minimal implementation. 
```
(which-key-mode 1)
```
And you will have something like this:
![image](https://user-images.githubusercontent.com/11796018/194339666-194c2a83-b5d8-4e30-a914-b73c938f31d4.png)

**Reply by aisamu** [reply]
(Unsure if you're replying to me, but just in case)

> You can use which-key with meow. 

I use it! 
To confirm: I re-tried the scenario posted on https://github.com/meow-edit/meow/issues/337#issuecomment-1268835642 after explicitly invoking `(which-key-mode 1)` and got the same results
